### PR TITLE
[Merged by Bors] - feat: Relation between `IsSuccLimit` and `iSup`.

### PR DESCRIPTION
**Pull request by erdOne**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/SuccPred/CompleteLinearOrder.lean [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+/-!
+
+# Relation between `IsSuccLimit` and `iSup` in (conditionally) complete linear orders.
+
+-/
+
+open Order
+
+variable {ι α : Type*}
+
```
Comment (urkud) [inline Mathlib/Order/SuccPred/CompleteLinearOrder.lean:18]: IMHO, the basic lemma should be something like
```
lemma sSup_mem_of_not_isSuccLimit [ConditionallyCompleteLinearOrder α]
    {s : Set α} (hne : s.Nonempty) (hbdd : BddAbove s) (hlim : ¬ IsSuccLimit x (sSup s)) :
    sSup s ∈ s := _
```

or even
```
lemma CovBy.sSup_mem [ConditionallyCompleteLinearOrder α] {a : α}
    {s : Set α} (hcov : a ⋖ sSup s) (hne : s.Nonempty) (hbdd : BddAbove s) :
    sSup s ∈ s := _
```

Also, please discuss relation between primed and non-primed versions in the module docstring, in lemma docstrings, or in both places.

File: Mathlib/Order/SuccPred/CompleteLinearOrder.lean [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+/-! ### ConditionallyCompleteLinearOrder -/
+
+lemma csSup_mem_of_not_isSuccLimit [ConditionallyCompleteLinearOrder α]
+    {s : Set α} (hne : s.Nonempty) (hbdd : BddAbove s) (hlim : ¬ IsSuccLimit (sSup s)) :
```
Comment (YaelDillies) [inline Mathlib/Order/SuccPred/CompleteLinearOrder.lean:22]:
```suggestion
section ConditionallyCompleteLinearOrder
variable [ConditionallyCompleteLinearOrder α] {s : Set α}

lemma csSup_mem_of_not_isSuccLimit (hne : s.Nonempty) (hbdd : BddAbove s)
    (hlim : ¬ IsSuccLimit (sSup s)) :
```

File: Mathlib/Order/SuccPred/CompleteLinearOrder.lean [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+# Relation between `IsSuccLimit` and `iSup` in (conditionally) complete linear orders.
+
+-/
+
+open Order
+
+variable {ι α : Type*}
+
+/-! ### ConditionallyCompleteLinearOrder -/
```
Comment (YaelDillies) [inline Mathlib/Order/SuccPred/CompleteLinearOrder.lean:19]: Can you either give a meaningful name to this subheading (not a Lean declaration name) or drop it? (that's why I was removing it in my previous comment)

Comment (erdOne) [reply Mathlib/Order/SuccPred/CompleteLinearOrder.lean:19]: I thought this is common practice to put lean declaration names in subheadings? Anyway I can trop it if it is unnecessary. I just wanted a place to put the docstring about primed names.

Comment (YaelDillies) [reply Mathlib/Order/SuccPred/CompleteLinearOrder.lean]: In subheadings? Not as a Lean declaration name, or at the very least in quotes, like
```
/-! ### `ConditionallyCompleteLinearOrder` -/
```

File: Mathlib/Order/SuccPred/CompleteLinearOrder.lean [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+lemma exists_of_not_isSuccLimit_ciSup [Nonempty ι]
+    (f : ι → α) (hf : BddAbove (Set.range f)) (hx : ¬ IsSuccLimit (⨆ i, f i)) :
+    ∃ i, f i = ⨆ i, f i :=
+  csSup_mem_of_not_isSuccLimit (Set.range_nonempty f) hf hx
+
+lemma exists_of_ciSup_eq_of_not_isSuccLimit [Nonempty ι]
+    (f : ι → α) (hf : BddAbove (Set.range f)) (x : α) (hx : ¬ IsSuccLimit x) (h : ⨆ i, f i = x) :
+    ∃ i, f i = x :=
+  h ▸ exists_of_not_isSuccLimit_ciSup f hf (h ▸ hx)
```
Comment (YaelDillies) [inline Mathlib/Order/SuccPred/CompleteLinearOrder.lean:37]: Those are better lemmas I would think
```suggestion
lemma IsLUB.mem_of_not_isSuccLimit (hx : ¬ IsSuccLimit x) (hs : IsLUB s x) : x ∈ s :=
  hs.csSup_eq ▸ csSup_mem_of_not_isSuccLimit hs.bddAbove (hs.csSup_eq ▸ hx)
  
lemma IsLUB.exists_of_not_isSuccLimit (hx : ¬ IsSuccLimit x) (hf : IsLUB (Set.range f) x) :
    ∃ i, f i = x := IsLUB.mem_of_not_isSuccLimit hx hf
```

File: Mathlib/Order/SuccPred/CompleteLinearOrder.lean (hunk 3 above)
Comment (erdOne) [reply Mathlib/Order/SuccPred/CompleteLinearOrder.lean]: Yeah maybe without the quotes. I removed the headings anyways

File: Mathlib/Order/SuccPred/CompleteLinearOrder.lean [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+# Relation between `IsSuccLimit` and `iSup` in (conditionally) complete linear orders.
+
+-/
+
+open Order
+
+variable {ι α : Type*}
+
+section ConditionallyCompleteLinearOrder
+variable [ConditionallyCompleteLinearOrder α] {s : Set α}
```
Comment (YaelDillies) [inline Mathlib/Order/SuccPred/CompleteLinearOrder.lean:20]:
```suggestion
variable [ConditionallyCompleteLinearOrder α] [Nonempty ι] {f : ι → α} {s : Set α} {x : α}
```

File: Mathlib/Order/SuccPred/CompleteLinearOrder.lean [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+lemma IsLUB.mem_of_nonempty_of_not_isSuccLimit
+    (hs : IsLUB s x) (hne : s.Nonempty) (hx : ¬ IsSuccLimit x) :
+    x ∈ s :=
+  hs.csSup_eq hne ▸ csSup_mem_of_not_isSuccLimit hne hs.bddAbove (hs.csSup_eq hne ▸ hx)
+
+lemma IsLUB.exists_of_nonempty_of_not_isSuccLimit
+    (hf : IsLUB (Set.range f) x) (hx : ¬ IsSuccLimit x) :
+    ∃ i, f i = x := hf.mem_of_nonempty_of_not_isSuccLimit (Set.range_nonempty f) hx
+
+lemma csInf_mem_of_not_isSuccLimit
```
Comment (YaelDillies) [inline Mathlib/Order/SuccPred/CompleteLinearOrder.lean:43]:
```suggestion
lemma csInf_mem_of_not_isPredLimit
```

File: Mathlib/Order/SuccPred/CompleteLinearOrder.lean [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+lemma exists_eq_ciInf_of_not_isPredLimit
+    (hf : BddBelow (Set.range f)) (hf' : ¬ IsPredLimit (⨅ i, f i)) :
+    ∃ i, f i = ⨅ i, f i :=
+  csInf_mem_of_not_isSuccLimit (Set.range_nonempty f) hf hf'
+
+lemma IsGLB.mem_of_nonempty_of_not_isPredLimit
+    (hs : IsGLB s x) (hne : s.Nonempty) (hx : ¬ IsPredLimit x) : x ∈ s :=
+  hs.csInf_eq hne ▸ csInf_mem_of_not_isSuccLimit hne hs.bddBelow (hs.csInf_eq hne ▸ hx)
+
+lemma IsGLB.exists_of_nonempty_not_isPredLimit
```
Comment (YaelDillies) [inline Mathlib/Order/SuccPred/CompleteLinearOrder.lean:59]:
```suggestion
lemma IsGLB.exists_of_nonempty_of_not_isPredLimit
```

File: Mathlib/Order/SuccPred/CompleteLinearOrder.lean [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    sSup s ∈ s := by
+  obtain ⟨y, hy⟩ := not_forall_not.mp hlim
+  obtain ⟨i, his, hi⟩ := lt_sSup_iff.mp hy.lt
+  exact eq_of_le_of_not_lt (le_sSup his) (hy.2 hi) ▸ his
+
+lemma exists_of_not_isSuccLimit_iSup (hf : ¬ IsSuccLimit (⨆ i, f i)) :
+    ∃ i, f i = ⨆ i, f i :=
+  sSup_mem_of_not_isSuccLimit hf
+
+lemma sInf_mem_of_not_isSuccLimit (hlim : ¬ IsPredLimit (sInf s)) :
```
Comment (YaelDillies) [inline Mathlib/Order/SuccPred/CompleteLinearOrder.lean:107]:
```suggestion
lemma sInf_mem_of_not_isPredLimit (hlim : ¬ IsPredLimit (sInf s)) :
```

**Review by YaelDillies — COMMENTED** [review]
Would you mind putting dual results next to each other? So the order would look like 
* `csSup_mem_of_not_isSuccLimit`
* `csInf_mem_of_not_isPredLimit`
* `exists_of_not_isSuccLimit_ciSup`
* `exists_of_not_isPredLimit_ciInf`

etc

File: Mathlib/Order/SuccPred/CompleteLinearOrder.lean [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+  csInf_mem_of_not_isPredLimit (Set.range_nonempty f) hf hf'
+
+lemma IsLUB.mem_of_nonempty_of_not_isSuccLimit
+    (hs : IsLUB s x) (hne : s.Nonempty) (hx : ¬ IsSuccLimit x) :
+    x ∈ s :=
+  hs.csSup_eq hne ▸ csSup_mem_of_not_isSuccLimit hne hs.bddAbove (hs.csSup_eq hne ▸ hx)
+
+lemma IsGLB.mem_of_nonempty_of_not_isPredLimit
+    (hs : IsGLB s x) (hne : s.Nonempty) (hx : ¬ IsPredLimit x) : x ∈ s :=
+  hs.csInf_eq hne ▸ csInf_mem_of_not_isPredLimit hne hs.bddBelow (hs.csInf_eq hne ▸ hx)
```
Comment (YaelDillies) [inline Mathlib/Order/SuccPred/CompleteLinearOrder.lean:53]: Can you make sure similar lemmas are similarly formatted?

File: Mathlib/SetTheory/Cardinal/Basic.lean [hunk 10]
```diff
@@ -1448,13 +1464,18 @@ theorem isLimit_aleph0 : IsLimit ℵ₀ :=
...
-  rcases lt_aleph0.1 h' with ⟨_ | n, rfl⟩
-  · exact h.ne_zero.irrefl
-  · rw [nat_succ] at h
-    exact not_isSuccLimit_succ _ h.isSuccLimit
-#align cardinal.is_limit.aleph_0_le Cardinal.IsLimit.aleph0_le
+  rcases lt_aleph0.1 h' with ⟨n, rfl⟩
+  exact not_isLimit_natCast n h
+
+lemma exists_of_iSup_eq_natCast {ι : Type u} [Nonempty ι] (f : ι → Cardinal.{max u v})
+    (n : ℕ) (h : ⨆ i : ι, f i = n) : ∃ i, f i = n :=
```
Comment (YaelDillies) [inline Mathlib/SetTheory/Cardinal/Basic.lean:1477]:
```suggestion
    (n : ℕ) (h : ⨆ i, f i = n) : ∃ i, f i = n :=
```

File: Mathlib/SetTheory/Cardinal/Basic.lean [hunk 11]
```diff
@@ -1448,13 +1464,18 @@ theorem isLimit_aleph0 : IsLimit ℵ₀ :=
...
   by_contra! h'
-  rcases lt_aleph0.1 h' with ⟨_ | n, rfl⟩
-  · exact h.ne_zero.irrefl
-  · rw [nat_succ] at h
-    exact not_isSuccLimit_succ _ h.isSuccLimit
-#align cardinal.is_limit.aleph_0_le Cardinal.IsLimit.aleph0_le
+  rcases lt_aleph0.1 h' with ⟨n, rfl⟩
+  exact not_isLimit_natCast n h
+
+lemma exists_of_iSup_eq_natCast {ι : Type u} [Nonempty ι] (f : ι → Cardinal.{max u v})
```
Comment (YaelDillies) [inline Mathlib/SetTheory/Cardinal/Basic.lean:1476]: Maybe this is a clearer name?
```suggestion
lemma exists_eq_natCast_of_iSup {ι : Type u} [Nonempty ι] (f : ι → Cardinal.{max u v})
```

**Review by YaelDillies — APPROVED** [review]
Thanks!

maintainer merge

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**